pkg/plugins/manager/loader/finder/fs.go

```diff
@@ -0,0 +1,293 @@
+package finder
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/fs"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+var walk = util.Walk
+
+var (
+	ErrInvalidPluginJSON         = errors.New("did not find valid type or id properties in plugin.json")
+	ErrInvalidPluginJSONFilePath = errors.New("invalid plugin.json filepath was provided")
+)
+
+type FS struct {
```

Comment:
Looks good, however I would separate `readPluginJSON` and `validatePluginJSON`  from `FS` as they contain non-FS related logic (like sanitization/validation etc).

For example, I would make `readPluginJSON` accept an `io.Reader` instead, so it can be used with non-FS as well.

I guess this was done like this for simplicity's sake (because we only support finding/loading from filesystem right now) and will be refactored once we prepare for CDN phase 2? If so, LGTM for the scope of this PR and we can do it for a follow-up PR along with CDN phase 2 :smile::+1:

Reply:
Aye good observation. Will save this one like you say!